### DISCUSSION TITLE
Расчёт дата + рабочие дни.

### DISCUSSION THREAD

**Discussion by EKhorikov** [totumonline/totum-ru-issues-and-discussions#62] [Как это сделать?] [answered]
Добрый день.
Вопрос по быстрым переменным - в курсе приведён пример с расчётом даты + 10 дней, через быструю переменную нынешней даты $#nd
= : dateAdd ( date: $#nd; days: 10; format: .....
А как можно рассчитать дату от сегодняшней, но + например 10 РАБОЧИХ дней? Это актуально в учёте договоров на какие-то услуги, где рассчитывается срок договора в рабочих днях. Например 40 рабочих дней с даты заключения.
Спасибо.

**Comment by ttm-alexeypolunin** [comment, marked as answer]
Тут можно пойти двумя путями.

### Первый

Мы выходными считаем только субботу и воскресенье, те 6 и 7 день недели:

```
= : while(preaction: $var_zero; action: $replace; postaction: $#date)

	var_zero: var(name: "date"; default: $#lc)

	replace: listReplace(list: $list; action: $var; key: "key")

			~list: listRepeat(item: 1; num: 40)
		
		var: var(name: "date"; value: $listadd)
			listadd: listAdd(list: $#date; item: $date)
				date: if(condition: $#key = 0; then: $#nd; else: $calc)
					calc: if(condition: $six_day = 6; then: $monday; else: $newdate)
						six_day: dateFormat(date: $newdate; format: "N")
						monday: dateAdd(date: $newdate; days: 2; format: "Y-m-d")
						newdate: dateAdd(date: $prev; days: 1; format: "Y-m-d")
							prev: $#date[$key]
								key: $#key - 1
```

В строке `list` параметр `num` замените на нужное вам количество дней.

В строке `date` — `$#nd` на стартовую дату.

### Второй

С учетом производственного календаря. Тогда нам нужна таблица в которой будут отмечены выходные дни. Обычно создается таблица циклов, где каждый цикл это год.

Внутри цикла список дат можно создать вот таким кодом-действия по кнопке:

```
= : while(preaction: $zero; preaction: $create_list; action: $insert)

	zero: var(name: "list"; value: $#lc)

create_list: while(condition: $getdate <= $enddate; action: $setvar; action: $setdate; limit: 380)
	getdate: var(name: "date"; default: $default)
		default: dateadd(date: $startdate; days: -1)
			~startdate: #h_start_year
	enddate: dateFormat(date: $startdate; format: "Y-12-31")

	setvar: var(name: "list"; value: $listadd)
		listadd: listadd(list: $#list; item: $dateadd)
			dateadd: dateadd(date: $#date; days: 1)
	
	setdate: var(name: "date"; value: $dateadd)

insert: insertlist(table: $#ntn; field: 'date' = $getlist)
getlist: var(name: "list")

// in [h_start_year] first date of the year
```

Дальше создаем там колонку `holiday` и определяем выходная дата по СБ и ВСК или нет:

```
= : if(condition: $date = json`[6,7]`; then: true; else: false)

date: dateFormat(date: #date; format: "N")
```

Дальше открываем производственный календарь на год, и ставим руками изменения где праздники и переносы.

> Если кто знает API по которому можно проверить статус даты производственного календаря в РФ — напишите в комментах.

Дальше мы проверяем по этой таблице, выходной ли день. И если выходной берем первый рабочий за ним:

```
= : while(preaction: $var_zero; action: $replace; postaction: $#date)

	var_zero: var(name: "date"; default: $#lc)

	replace: listReplace(list: $list; action: $var; key: "key")

			~list: listRepeat(item: 1; num: 40)
		
		var: var(name: "date"; value: $listadd)
			listadd: listAdd(list: $#date; item: $date)
				date: if(condition: $#key = 0; then: $#nd; else: $calc)
					calc: if(condition: $holiday = true; then: $next_work_day; else: $newdate)
						holiday: select(table: 'dates'; cycle: #year; field: 'holyday'; where: 'date' = $newdate)
						next_work_day: select(table: 'dates'; cycle: #year; field: 'date'; where: 'date' > $newdate; where: 'holyday' = false; order: 'date' asc)
						newdate: dateAdd(date: $prev; days: 1; format: "Y-m-d")
							prev: $#date[$key]
								key: $#key - 1

// replace [table] and [cycle] in rows [holiday] and [next_work_day]
```

**Comment by erpic-ru** [comment]
API производственного календаря: https://isdayoff.ru/api/getdata?year=2022

**Comment by Alexmeb** [comment]
Хотелось бы законченного решения с учетом api, и наверное с каким то примером. Вопрос насущный, как говорится)

**Reply by ttm-alexeypolunin** [reply]
Ок, 3-4 дня
